File: debop4s-timeperiod/src/test/scala/debop4s/timeperiod/tests/timeranges/HalfyearRangeTest.scala
package debop4s.timeperiod.tests.timeranges

import debop4s.core.parallels.Parallels
import debop4s.timeperiod._
import debop4s.timeperiod.tests.AbstractTimePeriodTest
import debop4s.timeperiod.timerange.HalfyearRange
import debop4s.timeperiod.utils.Times
import debop4s.timeperiod.utils.Times._

/**
 * HalfyearRangeTest
 * Created by debop on 2014. 2. 16.
 */
class HalfyearRangeTest extends AbstractTimePeriodTest {

  val now = Times.now
  val currentYear = now.getYear
  val calendar = EmptyOffsetTimeCalendar

  test("init values") {

    val firstHalfyear = startTimeOfHalfyear(now.getYear, Halfyear.First)
    val secondHalfyear = startTimeOfHalfyear(now.getYear, Halfyear.Second)

    val hyr = HalfyearRange(now.getYear, Halfyear.First, EmptyOffsetTimeCalendar)

    hyr.start.year shouldEqual firstHalfyear.year
    hyr.start.getMonthOfYear shouldEqual firstHalfyear.getMonthOfYear
    hyr.start.getDayOfMonth shouldEqual firstHalfyear.getDayOfMonth
    hyr.start.getHourOfDay shouldEqual 0
    hyr.start.getMinuteOfHour shouldEqual 0
    hyr.start.getSecondOfMinute shouldEqual 0
    hyr.start.getMillisOfSecond shouldEqual 0

    hyr.end.year shouldEqual secondHalfyear.year
    hyr.end.getMonthOfYear shouldEqual secondHalfyear.getMonthOfYear
    hyr.end.getDayOfMonth shouldEqual secondHalfyear.getDayOfMonth
    hyr.end.getHourOfDay shouldEqual 0
    hyr.end.getMinuteOfHour shouldEqual 0
    hyr.end.getSecondOfMinute shouldEqual 0
    hyr.end.getMillisOfSecond shouldEqual 0
  }

  test("default calendar") {
    val yearStart = startTimeOfYear(currentYear)
    for (halfyear <- Halfyear.values) {
      val offset = halfyear.id - 1
      val hyr = HalfyearRange(yearStart.plusMonths(MonthsPerHalfyear * offset))

      hyr.start shouldEqual hyr.calendar.mapStart(yearStart.plusMonths(MonthsPerHalfyear * offset))
      hyr.end shouldEqual hyr.calendar.mapEnd(yearStart.plusMonths(MonthsPerHalfyear * ( offset + 1 )))
    }
  }

  test("moment") {
    HalfyearRange().halfyear shouldEqual ( if (now.getMonthOfYear < 7) Halfyear.First else Halfyear.Second )

    Parallels.runAction1(MonthsPerYear)(m => {
      val month = m + 1
      HalfyearRange(currentYear, Halfyear.First).startMonthOfYear shouldEqual 1
      HalfyearRange(currentYear, Halfyear.Second).startMonthOfYear shouldEqual 7
    })
  }

  test("multiple calendar years") {
    val currentYear = now.getYear
    HalfyearRange(currentYear, Halfyear.First).isMultipleCalendarYears shouldEqual false
  }

  test("calendar halfyear") {

    val h1 = HalfyearRange(currentYear, Halfyear.First, calendar)

    h1.isReadonly shouldEqual true
    h1.halfyear shouldEqual Halfyear.First
    h1.start shouldEqual asDate(currentYear, 1, 1)
    h1.end shouldEqual asDate(currentYear, 7, 1)

    val h2 = HalfyearRange(currentYear, Halfyear.Second, calendar)

    h2.isReadonly shouldEqual true
    h2.halfyear shouldEqual Halfyear.Second
    h2.start shouldEqual asDate(currentYear, 7, 1)
    h2.end shouldEqual asDate(currentYear + 1, 1, 1)
  }

  test("halfyear quaters") {

    val h1 = HalfyearRange(currentYear, Halfyear.First, calendar)
    val h1quarters = h1.getQuarters

    var h1index = 0
    h1quarters.foreach { qr =>
      log.trace(s"qr=$qr")
      qr.quarter shouldEqual ( if (h1index == 0) Quarter.First else Quarter.Second )
      qr.start shouldEqual h1.start.plusMonths(h1index * MonthsPerQuarter)
      qr.end shouldEqual h1.calendar.mapEnd(qr.start.plusMonths(MonthsPerQuarter))
      h1index += 1
    }

    val h2 = HalfyearRange(currentYear, Halfyear.Second, calendar)
    val h2quarters = h2.getQuarters

    var h2index = 0
    h2quarters.foreach { qr =>
      log.trace(s"qr=$qr")
      qr.quarter shouldEqual ( if (h2index == 0) Quarter.Third else Quarter.Fourth )
      qr.start shouldEqual h2.start.plusMonths(h2index * MonthsPerQuarter)
      qr.end shouldEqual qr.calendar.mapEnd(qr.start.plusMonths(MonthsPerQuarter))
      h2index += 1
    }
  }

  test("halfyear getMonths") {

    val h1 = HalfyearRange(currentYear, Halfyear.First, calendar)
    val months = h1.getMonths
    months.size shouldEqual MonthsPerHalfyear

    var index = 0
    months.foreach { m =>
      m.start shouldEqual h1.start.plusMonths(index)
      m.end shouldEqual calendar.mapEnd(m.start.plusMonths(1))
      index += 1
    }
  }

  test("add halfyears") {

    val h1 = HalfyearRange(currentYear, Halfyear.First, calendar)

    var prevH1 = h1.addHalfyears(-1)
    prevH1.halfyear shouldEqual Halfyear.Second
    prevH1.year shouldEqual ( currentYear - 1 )
    prevH1.start shouldEqual h1.start.plusMonths(-MonthsPerHalfyear)
    prevH1.end shouldEqual h1.start

    prevH1 = h1.addHalfyears(-2)
    prevH1.halfyear shouldEqual Halfyear.First
    prevH1.year shouldEqual ( currentYear - 1 )
    prevH1.start shouldEqual h1.start.plusMonths(-2 * MonthsPerHalfyear)
    prevH1.end shouldEqual h1.start.plusMonths(-1 * MonthsPerHalfyear)

    prevH1 = h1.addHalfyears(-3)
    prevH1.halfyear shouldEqual Halfyear.Second
    prevH1.year shouldEqual ( currentYear - 2 )
    prevH1.start shouldEqual h1.start.plusMonths(-3 * MonthsPerHalfyear)
    prevH1.end shouldEqual h1.start.plusMonths(-2 * MonthsPerHalfyear)

    var nextH1 = h1.addHalfyears(1)
    nextH1.halfyear shouldEqual Halfyear.Second
    nextH1.year shouldEqual currentYear
    nextH1.start shouldEqual h1.start.plusMonths(MonthsPerHalfyear)
    nextH1.end shouldEqual h1.start.plusMonths(2 * MonthsPerHalfyear)

    nextH1 = h1.addHalfyears(2)
    nextH1.halfyear shouldEqual Halfyear.First
    nextH1.year shouldEqual ( currentYear + 1 )
    nextH1.start shouldEqual h1.start.plusMonths(2 * MonthsPerHalfyear)
    nextH1.end shouldEqual h1.start.plusMonths(3 * MonthsPerHalfyear)

    nextH1 = h1.addHalfyears(3)
    nextH1.halfyear shouldEqual Halfyear.Second
    nextH1.year shouldEqual ( currentYear + 1 )
    nextH1.start shouldEqual h1.start.plusMonths(3 * MonthsPerHalfyear)
    nextH1.end shouldEqual h1.start.plusMonths(4 * MonthsPerHalfyear)
  }

}

File: debop4s-timeperiod/src/test/java/debop4s/timeperiod/tests/timeranges/HalfyearRangeTest.java
package debop4s.timeperiod.tests.timeranges;

import debop4s.core.JAction1;
import debop4s.core.parallels.JParallels;
import debop4s.timeperiod.Halfyear;
import debop4s.timeperiod.Quarter;
import debop4s.timeperiod.TimeCalendar;
import debop4s.timeperiod.TimeSpec;
import debop4s.timeperiod.tests.TimePeriodTestBase;
import debop4s.timeperiod.timerange.HalfyearRange;
import debop4s.timeperiod.timerange.MonthRange;
import debop4s.timeperiod.timerange.QuarterRange;
import debop4s.timeperiod.utils.Times;
import lombok.extern.slf4j.Slf4j;
import org.joda.time.DateTime;
import org.junit.Test;

import java.util.List;

import static org.fest.assertions.Assertions.assertThat;


@Slf4j
public class HalfyearRangeTest extends TimePeriodTestBase {

    @Test
    public void initValues() {
        DateTime now = Times.now();
        DateTime firstHalfyear = Times.startTimeOfHalfyear(now.getYear(), Halfyear.First);
        DateTime secondHalfyear = Times.startTimeOfHalfyear(now.getYear(), Halfyear.Second);

        HalfyearRange hyr = new HalfyearRange(now.getYear(), Halfyear.First, TimeCalendar.getEmptyOffset());

        assertThat(hyr.getStart().getYear()).isEqualTo(firstHalfyear.getYear());
        assertThat(hyr.getStart().getMonthOfYear()).isEqualTo(firstHalfyear.getMonthOfYear());
        assertThat(hyr.getStart().getDayOfMonth()).isEqualTo(firstHalfyear.getDayOfMonth());
        assertThat(hyr.getStart().getHourOfDay()).isEqualTo(0);
        assertThat(hyr.getStart().getMinuteOfHour()).isEqualTo(0);
        assertThat(hyr.getStart().getMinuteOfHour()).isEqualTo(0);
        assertThat(hyr.getStart().getSecondOfMinute()).isEqualTo(0);
        assertThat(hyr.getStart().getMillisOfSecond()).isEqualTo(0);

        assertThat(hyr.getEnd().getYear()).isEqualTo(secondHalfyear.getYear());
        assertThat(hyr.getEnd().getMonthOfYear()).isEqualTo(secondHalfyear.getMonthOfYear());
        assertThat(hyr.getEnd().getDayOfMonth()).isEqualTo(secondHalfyear.getDayOfMonth());
        assertThat(hyr.getEnd().getHourOfDay()).isEqualTo(0);
        assertThat(hyr.getEnd().getMinuteOfHour()).isEqualTo(0);
        assertThat(hyr.getEnd().getMinuteOfHour()).isEqualTo(0);
        assertThat(hyr.getEnd().getSecondOfMinute()).isEqualTo(0);
        assertThat(hyr.getEnd().getMillisOfSecond()).isEqualTo(0);
    }

    @Test
    public void defaultCalendarTest() {
        DateTime yearStart = Times.startTimeOfYear(Times.currentYear());

        for (Halfyear halfyear : Halfyear.values()) {
            int offset = halfyear.getValue() - 1;
            HalfyearRange hyr = new HalfyearRange(yearStart.plusMonths(TimeSpec.MonthsPerHalfyear * offset));

            assertThat(hyr.getStart()).isEqualTo(hyr.getTimeCalendar().mapStart(yearStart.plusMonths(TimeSpec.MonthsPerHalfyear * offset)));
            assertThat(hyr.getEnd()).isEqualTo(hyr.getTimeCalendar().mapEnd(yearStart.plusMonths(TimeSpec.MonthsPerHalfyear * (offset + 1))));
        }
    }

    @Test
    public void momentTest() {
        DateTime now = Times.now();
        int currentYear = now.getYear();
        assertThat(new HalfyearRange().getHalfyear()).isEqualTo(
                now.getMonthOfYear() < 7 ? Halfyear.First : Halfyear.Second);

        assertThat(new HalfyearRange(Times.asDate(currentYear, 1, 1)).getHalfyear()).isEqualTo(Halfyear.First);
        assertThat(new HalfyearRange(Times.asDate(currentYear, 6, 30)).getHalfyear()).isEqualTo(Halfyear.First);
        assertThat(new HalfyearRange(Times.asDate(currentYear, 7, 1)).getHalfyear()).isEqualTo(Halfyear.Second);
        assertThat(new HalfyearRange(Times.asDate(currentYear, 12, 31)).getHalfyear()).isEqualTo(Halfyear.Second);
    }

    @Test
    public void startMonth() {
        final DateTime now = Times.now();
        final int currentYear = now.getYear();

        JParallels.run(0, TimeSpec.MonthsPerYear, new JAction1<Integer>() {
            @Override
            public void perform(Integer m) {
                int month = m + 1;
                assertThat(new HalfyearRange(currentYear, Halfyear.First).getStartMonthOfYear()).isEqualTo(1);
                assertThat(new HalfyearRange(currentYear, Halfyear.Second).getStartMonthOfYear()).isEqualTo(7);
            }
        });
    }

    @Test
    public void isMultipleCalendarYearsTest() {
        final DateTime now = Times.now();
        final int currentYear = now.getYear();

        assertThat(new HalfyearRange(currentYear, Halfyear.First).isMultipleCalendarYears()).isFalse();
    }

    @Test
    public void calendarHalfyear() {
        final DateTime now = Times.now();
        final int currentYear = now.getYear();
        final TimeCalendar calendar = TimeCalendar.getEmptyOffset();

        HalfyearRange h1 = new HalfyearRange(currentYear, Halfyear.First, calendar);

        assertThat(h1.isReadonly()).isTrue();
        assertThat(h1.getHalfyear()).isEqualTo(Halfyear.First);
        assertThat(h1.getStart()).isEqualTo(Times.asDate(currentYear, 1, 1));
        assertThat(h1.getEnd()).isEqualTo(Times.asDate(currentYear, 7, 1));

        HalfyearRange h2 = new HalfyearRange(currentYear, Halfyear.Second, calendar);

        assertThat(h2.isReadonly()).isTrue();
        assertThat(h2.getHalfyear()).isEqualTo(Halfyear.Second);
        assertThat(h2.getStart()).isEqualTo(Times.asDate(currentYear, 7, 1));
        assertThat(h2.getEnd()).isEqualTo(Times.asDate(currentYear + 1, 1, 1));
    }

    @Test
    public void getQuartersTest() {
        final DateTime now = Times.now();
        final int currentYear = now.getYear();
        final TimeCalendar calendar = TimeCalendar.getEmptyOffset();

        HalfyearRange h1 = new HalfyearRange(currentYear, Halfyear.First, calendar);
        List<QuarterRange> h1Quarters = h1.quarters();

        int h1Index = 0;
        for (QuarterRange h1Quarter : h1Quarters) {
            log.trace("h1Quarter[{}] = [{}]", h1Index, h1Quarter);
            assertThat(h1Quarter.getQuarter()).isEqualTo(h1Index == 0 ? Quarter.First : Quarter.Second);
            assertThat(h1Quarter.getStart()).isEqualTo(h1.getStart().plusMonths(h1Index * TimeSpec.MonthsPerQuarter));
            assertThat(h1Quarter.getEnd()).isEqualTo(h1Quarter.getTimeCalendar().mapEnd(h1Quarter.getStart().plusMonths(TimeSpec.MonthsPerQuarter)));
            h1Index++;
        }

        HalfyearRange h2 = new HalfyearRange(currentYear, Halfyear.Second, calendar);
        List<QuarterRange> h2Quarters = h2.quarters();

        int h2Index = 0;
        for (QuarterRange h2Quarter : h2Quarters) {
            log.trace("h2Quarter[{}] = [{}]", h2Index, h2Quarter);
            assertThat(h2Quarter.getQuarter()).isEqualTo(h2Index == 0 ? Quarter.Third : Quarter.Fourth);
            assertThat(h2Quarter.getStart()).isEqualTo(h2.getStart().plusMonths(h2Index * TimeSpec.MonthsPerQuarter));
            assertThat(h2Quarter.getEnd()).isEqualTo(h2Quarter.getTimeCalendar().mapEnd(h2Quarter.getStart().plusMonths(TimeSpec.MonthsPerQuarter)));
            h2Index++;
        }
    }

    @Test
    public void getMonthsTest() {
        final DateTime now = Times.now();
        final int currentYear = now.getYear();
        final TimeCalendar calendar = TimeCalendar.getEmptyOffset();

        HalfyearRange h1 = new HalfyearRange(currentYear, Halfyear.First, calendar);
        List<MonthRange> months = h1.months();
        assertThat(months.size()).isEqualTo(TimeSpec.MonthsPerHalfyear);

        int index = 0;
        for (MonthRange month : months) {
            assertThat(month.getStart()).isEqualTo(h1.getStart().plusMonths(index));
            assertThat(month.getEnd()).isEqualTo(calendar.mapEnd(month.getStart().plusMonths(1)));
            index++;
        }
    }

    @Test
    public void addHalfyearsTest() {
        final DateTime now = Times.now();
        final int currentYear = now.getYear();
        final TimeCalendar calendar = TimeCalendar.getEmptyOffset();

        HalfyearRange h1 = new HalfyearRange(currentYear, Halfyear.First, calendar);

        HalfyearRange prevH1 = h1.addHalfyears(-1);
        assertThat(prevH1.getHalfyear()).isEqualTo(Halfyear.Second);
        assertThat(prevH1.getStart()).isEqualTo(h1.getStart().plusMonths(-TimeSpec.MonthsPerHalfyear));
        assertThat(prevH1.getEnd()).isEqualTo(h1.getStart());

        prevH1 = h1.addHalfyears(-2);
        assertThat(prevH1.getHalfyear()).isEqualTo(Halfyear.First);
        assertThat(prevH1.getStart()).isEqualTo(h1.getStart().plusMonths(-2 * TimeSpec.MonthsPerHalfyear));
        assertThat(prevH1.getEnd()).isEqualTo(h1.getStart().plusMonths(-1 * TimeSpec.MonthsPerHalfyear));

        prevH1 = h1.addHalfyears(-3);
        assertThat(prevH1.getHalfyear()).isEqualTo(Halfyear.Second);
        assertThat(prevH1.getStart()).isEqualTo(h1.getStart().plusMonths(-3 * TimeSpec.MonthsPerHalfyear));
        assertThat(prevH1.getEnd()).isEqualTo(h1.getStart().plusMonths(-2 * TimeSpec.MonthsPerHalfyear));

        HalfyearRange nextH1 = h1.addHalfyears(1);
        assertThat(nextH1.getHalfyear()).isEqualTo(Halfyear.Second);
        assertThat(nextH1.getStart()).isEqualTo(h1.getStart().plusMonths(TimeSpec.MonthsPerHalfyear));
        assertThat(nextH1.getEnd()).isEqualTo(h1.getStart().plusMonths(2 * TimeSpec.MonthsPerHalfyear));

        nextH1 = h1.addHalfyears(2);
        assertThat(nextH1.getHalfyear()).isEqualTo(Halfyear.First);
        assertThat(nextH1.getStart()).isEqualTo(h1.getStart().plusMonths(2 * TimeSpec.MonthsPerHalfyear));
        assertThat(nextH1.getEnd()).isEqualTo(h1.getStart().plusMonths(3 * TimeSpec.MonthsPerHalfyear));

        nextH1 = h1.addHalfyears(3);
        assertThat(nextH1.getHalfyear()).isEqualTo(Halfyear.Second);
        assertThat(nextH1.getStart()).isEqualTo(h1.getStart().plusMonths(3 * TimeSpec.MonthsPerHalfyear));
        assertThat(nextH1.getEnd()).isEqualTo(h1.getStart().plusMonths(4 * TimeSpec.MonthsPerHalfyear));

    }
}
